src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs

```diff
@@ -17,12 +18,134 @@ internal static class TestCommandParser
     {
         public static readonly string DocsLink = "https://aka.ms/dotnet-test";
 
-        public static readonly Argument<string> SlnOrProjectArgument = new Argument<string>(CommonLocalizableStrings.SolutionOrProjectArgumentName)
+        /// <summary>
+        /// Parser delegate that only accepts a token that is a 
+        /// * project or solution file
+        /// * directory that contains a project or solution
+        /// * .dll or .exe file
+        /// </summary>
+        /// <remarks>
+        /// S.CL usage note - OnlyTake(0) signals that this token should be returned to the
+        /// token stream for the next argument. In this way we prevent initial tokens that 
+        /// are not relevant from being captured in this argument, since the syntax is a bit
+        /// ambiguous. 
+        /// </remarks>
+        private static ParseArgument<string> IsProjectOrSln =
+            (ctx) =>
+            {
+                bool HasProjectOrSolution(DirectoryInfo dir) =>
+                    dir.EnumerateFiles("*.*proj").Any() || dir.EnumerateFiles(".sln").Any();
+
+                if (ctx.Tokens.Count == 0)
+                {
+                    ctx.OnlyTake(0);
+                    return null;
+                }
+                else
+                {
+                    var tokenValue = ctx.Tokens[0].Value;
+                    var ext = System.IO.Path.GetExtension(tokenValue);
+                    if (ext.EndsWith("proj") || ext.EndsWith(".sln") || ext.EndsWith(".dll") || ext.EndsWith(".exe"))
+                    {
+                        ctx.OnlyTake(1);
+                        return tokenValue;
+                    }
+                    else
+                    {
+                        var path = System.IO.Path.GetFullPath(tokenValue);
+                        var dir = new System.IO.DirectoryInfo(path);
+                        if (dir.Exists && HasProjectOrSolution(dir))
+                        {
+                            ctx.OnlyTake(1);
+                            return tokenValue;
+                        }
+
+                        ctx.OnlyTake(0);
+                        return null;
```

Comment:
Can we return error to user? If I understand correct now we will just run like argument has never passed.

Reply:
we don't want to return an error to the user in this case - the test runner has specific logic to handle a sln/proj/dll/exe not being specified and if it ends up being a _real_ problem their error reporting is preferable IMO. With these arguments I'm trying to minimally-model their command line to reduce ambiguity only, any actual processing should be done on their end.

Reply:
it's very possible for this argument to _not_ be passed, per the docs we have for dotnet test. so I'm attempting to be very defensive here and only bind this argument if it's valid, otherwise let the token be parsed as a forwarded argument or a runsetting.